africanus/dft/kernels.py

```diff
@@ -76,30 +75,40 @@ def vis_to_im(vis, uvw, lm, frequency, dtype=None):
 
         out_dtype = dtype.dtype
 
-    def _vis_to_im_impl(vis, uvw, lm, frequency, dtype=None):
-        im_of_vis = np.zeros((lm.shape[0], frequency.shape[0]),
-                             dtype=out_dtype)
+    assert np.shape(vis) == np.shape(flags)
+
+    def _vis_to_im_impl(vis, uvw, lm, frequency, flags, dtype=None):
+        nrows = uvw.shape[0]
+        nsrc = lm.shape[0]
+        nchan = frequency.shape[0]
+        ncorr = vis.shape[-1]
+
+        im_of_vis = np.zeros((nsrc, nchan, ncorr), dtype=out_dtype)
 
         # For each source
-        for source in range(lm.shape[0]):
-            l, m = lm[source]
+        for s in range(nsrc):
+            l, m = lm[s]
             n = np.sqrt(1.0 - l ** 2 - m ** 2) - 1.0
             # For each uvw coordinate
-            for row in range(uvw.shape[0]):
-                u, v, w = uvw[row]
+            for r in range(nrows):
+                u, v, w = uvw[r]
 
                 # e^(-2*pi*(l*u + m*v + n*w)/c)
                 real_phase = -minus_two_pi_over_c * (l * u + m * v + n * w)
 
                 # Multiple in frequency for each channel
-                for chan in range(frequency.shape[0]):
-                    p = real_phase * frequency[chan]
-
-                    im_of_vis[source,
-                              chan] += (np.cos(p) * vis[row, chan].real -
-                                        np.sin(p) * vis[row, chan].imag)
-                    # Note for the adjoint we don't need the imaginary part
-                    # and we can elide the call to exp
+                for nu in range(nchan):
+                    p = real_phase * frequency[nu]
+
+                    # do not compute if any of the correlations
+                    # are flagged (complicates uncertainties)
+                    if not np.any(flags[r, nu]):
```

Comment:
Consider changing this to the following as it gives you more whitespace to play with

```python
if np.any(flags[r, nu]):
    continue

for c in range(ncorr):
   ...
```

Reply:
crafty, thanks